### PR TITLE
Shard eval dataset and aggregate eval metrics

### PR DESCRIPTION
**Pull request by jysohn23**
Also, instead of calling `eval_loss.item()` every time do summation with
tensors on device.

### CODE REVIEW

File: examples/run_glue_tpu.py [hunk 1]
```diff
@@ -256,14 +257,29 @@ def evaluate(args, model, tokenizer, prefix="", disable_logging=False):
             preds = np.squeeze(preds)
         result = compute_metrics(eval_task, preds, out_label_ids)
         results.update(result)
+        results['eval_loss'] = eval_loss.item()
+
+        # Average all metrics from each shard
```
Comment (taylanbil) [inline examples/run_glue_tpu.py]: what are some of the metrics? does it make sense to avg all of them? some metrics may be additive.

File: examples/run_glue_tpu.py [hunk 2]
```diff
@@ -256,14 +257,29 @@ def evaluate(args, model, tokenizer, prefix="", disable_logging=False):
...
+            'eval_avg_metrics', results, reduce_fn)
 
         output_eval_file = os.path.join(eval_output_dir, prefix, "eval_results.txt")
-        with open(output_eval_file, "w") as writer:
-            logger.info("***** Eval results {} *****".format(prefix))
-            for key in sorted(result.keys()):
-                logger.info("  %s = %s", key, str(result[key]))
-                writer.write("%s = %s\n" % (key, str(result[key])))
-                tb_writer.add_scalar(key, result[key])
+        if xm.is_master_ordinal():
```
Comment (taylanbil) [inline examples/run_glue_tpu.py:268]: is everything being logged here already on cpu?

File: examples/run_glue_tpu.py [hunk 3]
```diff
@@ -256,14 +257,29 @@ def evaluate(args, model, tokenizer, prefix="", disable_logging=False):
             preds = np.squeeze(preds)
         result = compute_metrics(eval_task, preds, out_label_ids)
         results.update(result)
+        results['eval_loss'] = eval_loss.item()
+
+        # Average all metrics from each shard
+        def reduce_fn(results):
+            avg_result = defaultdict(int)
```
Comment (taylanbil) [inline examples/run_glue_tpu.py]: why not `defaultdict(float)`?

**Review by taylanbil — COMMENTED** [review]
did you test this e2e?

File: examples/run_glue_tpu.py (hunk 1 above)
Comment (jysohn23) [reply examples/run_glue_tpu.py]: f1, accuracy, eval_loss, acc_and_f1 (avg of the two)

I checked they all make sense averaged.

File: examples/run_glue_tpu.py (hunk 3 above)
Comment (jysohn23) [reply examples/run_glue_tpu.py]: Good point. Done.

File: examples/run_glue_tpu.py (hunk 2 above)
Comment (jysohn23) [reply examples/run_glue_tpu.py:268]: Yes.

**Review by jysohn23 — COMMENTED** [review]
Yeah, e2e tested.

File: examples/run_glue_tpu.py (hunk 1 above)
Comment (jysohn23) [reply examples/run_glue_tpu.py]: Yep ignore that :D Not additive as discussed. Will update PR.

Comment (taylanbil) [reply examples/run_glue_tpu.py]: I'd be surprised if `global f1 == np.mean(local f1s)`. That's probably not true, let's verify on paper what the true formula to get global f1 is.

File: examples/run_glue_tpu.py (hunk 2 above)
Comment (taylanbil) [reply examples/run_glue_tpu.py:268]: maybe let's add a comment? It's a subtle point that can be missed by code readers.

File: examples/run_glue_tpu.py (hunk 1 above)
Comment (jysohn23) [reply examples/run_glue_tpu.py]: Updated to sync the pred/label tensors directly instead (these shouldn't be that big for finetuning tasks; single integer 0/1 per example). This way we don't have to have some custom aggregation per metric and don't touch upstream core code.

**Review by taylanbil — APPROVED** [review]

**Pull request merged**